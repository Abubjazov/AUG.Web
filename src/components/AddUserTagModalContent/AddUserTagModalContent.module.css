.root {
  display: flex;
  flex-direction: column;
  gap: 20px;
  width: 80vw;
  color: var(--black);
  animation: var(--animation-visible);
}

.title {
  font-size: 48px;
  font-weight: 900;
  line-height: 48px;
}

.text {
  font-size: 16px;
  font-weight: 500;
  line-height: 16px;
}

.buttons {
  display: flex;
  gap: 20px;
}

.button {
  height: 25px;
  margin-top: 20px;
}
