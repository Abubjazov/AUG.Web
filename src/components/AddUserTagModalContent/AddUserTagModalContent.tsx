import { FC } from 'react'

import TagsGroup from 'components/TagsGroup'
import { useAppDispatch, useAppSelector } from 'store/hooks'
import { setModalState } from 'store/slices/layoutSlice'
import BaseButton from 'uikit/BaseButton'
import { BaseButtonMode } from 'uikit/BaseButton/BaseButton'
import { SmartTagMode } from 'uikit/SmartTag/SmartTag'

import styles from './AddUserTagModalContent.module.css'

export enum ERenderMode {
  WELCOME = 'welcome',
  LOGIN = 'login',
  REGISTRATION = 'registration',
}

export interface AddUserTagModalContentProps {
  dappletId: string
}

const AddUserTagModalContent: FC<AddUserTagModalContentProps> = ({
  dappletId,
}) => {
  const dispatch = useAppDispatch()

  const { userTags, tagOperationGoing } = useAppSelector(
    state => state.userData,
  )

  return (
    <div className={styles['root']}>
      <TagsGroup
        menuOpened={true}
        tags={userTags}
        title={'My tags'}
        tagMode={SmartTagMode.MY_TAG_MODAL}
        titleUppercase
        tagOperationGoing={tagOperationGoing}
        dappletId={dappletId}
      />

      <div className={styles.buttons}>
        <BaseButton
          userStyles={styles.button}
          label={'Close'}
          mode={BaseButtonMode.CONTAINED_RED}
          onClick={() => dispatch(setModalState(false))}
        />
      </div>
    </div>
  )
}

export default AddUserTagModalContent
