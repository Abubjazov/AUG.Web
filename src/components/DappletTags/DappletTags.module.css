.root {
  display: flex;
  flex-wrap: wrap;
  gap: 8px;
}

.tag-root:hover {
  box-shadow: var(--box-shadow-2);
}

.button {
  height: 17px;
  padding: 0 15px;
  font-size: 15px;
  font-weight: 500;
  line-height: 13px;
  color: var(--white);
  cursor: pointer;
  background-color: var(--graylight);
  border-radius: 4px;
}
