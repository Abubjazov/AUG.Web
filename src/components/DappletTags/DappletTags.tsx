import { FC } from 'react'

import { useResize } from 'hooks/useResize/useResize'
import { nanoid } from 'nanoid'
import { removeUserTagFromDapplet } from 'services/userData/userData'
import { useAppDispatch, useAppSelector } from 'store/hooks'
import {
  EDappletOperation,
  IDappletOperation,
} from 'store/slices/userDataSlice'
import SmartTag from 'uikit/SmartTag'
import { SmartTagMode } from 'uikit/SmartTag/SmartTag'
import { combineClasses as cc } from 'utils/combineClasses/combineClasses'

import styles from './DappletTags.module.css'

export interface DappletTagsProps {
  userStyles?: string
  dappletId: string
  dappletState: boolean
  dappletOperationGoing?: IDappletOperation[]
}

const DappletTags: FC<DappletTagsProps> = ({
  userStyles = '',
  dappletId,
  dappletState,
  dappletOperationGoing = [],
}) => {
  const windowInnerWidth = useResize()

  const dispatch = useAppDispatch()

  const { dapplets, tags } = useAppSelector(state => state.dapplets)
  const { userDapplets, userTags } = useAppSelector(state => state.userData)

  const targetAllDapplet = dapplets.filter(
    dapplet => dapplet.dappletId === dappletId,
  )[0]

  const targetMyDapplets = userDapplets.filter(
    dapplet => dapplet.dappletId === dappletId,
  )[0]

  const onClickHandler = (userTagId: string) => {
    void dispatch(
      removeUserTagFromDapplet({
        dappletId,
        userTagId,
      }),
    )
  }

  return (
    <div className={cc([styles.root, userStyles])}>
      {targetMyDapplets &&
        targetMyDapplets.userTags.map(tagId => {
          const tagName = userTags.filter(tag => tag.tagId === tagId)[0]
            ?.tagName

          return (
            <SmartTag
              key={nanoid()}
              mode={SmartTagMode.MY_TAG}
              tagId={tagId}
              label={tagName}
              onClick={onClickHandler}
              loading={Boolean(
                dappletOperationGoing.filter(
                  operation =>
                    operation.dappletId === dappletId &&
                    operation.userTagId === tagId &&
                    operation.operation === EDappletOperation.REMOVE_USER_TAG,
                ).length,
              )}
            />
          )
        })}

      {targetAllDapplet &&
        targetAllDapplet.communityTags.map(tagId => {
          const tagName = tags.filter(tag => tag.tagId === tagId)[0]?.tagName

          return (
            <SmartTag
              key={nanoid()}
              mode={SmartTagMode.COMMUNITY_TAG}
              tagId={tagId}
              label={tagName}
            />
          )
        })}

      {windowInnerWidth <= 880 && dappletState && (
        <button
          type="button"
          data-testid="cross-button"
          className={styles.button}
        >
          +
        </button>
      )}
    </div>
  )
}

export default DappletTags
