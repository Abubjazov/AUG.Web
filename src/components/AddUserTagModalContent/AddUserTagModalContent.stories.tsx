import type { Meta, StoryObj } from '@storybook/react'
import MockedProvider from 'mockData/mockedReduxProvider'

import AddUserTagModalContent, {
  AddUserTagModalContentProps,
} from './AddUserTagModalContent'

import '/src/index.css'

const meta: Meta<AddUserTagModalContentProps> = {
  component: AddUserTagModalContent,
  title: 'COMPONENTS/Add User Tag Modal Content',
  parameters: {
    docs: {
      description: {
        component: 'Interface component - AddUserTagModalContent',
      },
    },
  },
  tags: ['autodocs'],
  argTypes: {
    dappletId: {
      description: 'Current dapplet ID',
    },
  },
}

export default meta

type Story = StoryObj<AddUserTagModalContentProps>

export const Default: Story = {
  args: {
    dappletId: 'ECNk2nNngwGXouvMpjWt',
  },

  decorators: [
    Story => (
      <MockedProvider>
        <Story />
      </MockedProvider>
    ),
  ],
}
