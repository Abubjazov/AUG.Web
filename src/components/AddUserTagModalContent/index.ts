import AddUserTagModalContent from './AddUserTagModalContent'

export default AddUserTagModalContent
