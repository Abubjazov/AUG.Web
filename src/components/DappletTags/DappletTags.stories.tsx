import type { Meta, StoryObj } from '@storybook/react'
import { mockedReduxProvider as Provider } from 'mockData/mockedReduxProvider'

import DappletTags, { DappletTagsProps } from './DappletTags'

import '/src/index.css'

const meta: Meta<DappletTagsProps> = {
  component: DappletTags,
  title: 'COMPONENTS/Dapplet Tags',
  parameters: {
    docs: {
      description: {
        component: 'Interface component - Dapplet Tags',
      },
    },
  },
  tags: ['autodocs'],
  argTypes: {
    userStyles: {
      description: 'Passing additional custom styles for root element',
    },
    dappletId: {
      description: 'Displayed dapplet ID',
    },
    dappletState: {
      description:
        'Dapplet state: open/close (only for Window Inner Width <= 880 )',
    },
  },
}

export default meta

type Story = StoryObj<DappletTagsProps>

export const Default: Story = {
  args: { dappletId: 1 },

  decorators: [
    Story => (
      <Provider>
        <div style={{ width: '300px' }}>
          <Story />
        </div>
      </Provider>
    ),
  ],
}
