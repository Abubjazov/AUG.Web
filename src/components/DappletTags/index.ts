import DappletTags from './DappletTags'

export default DappletTags
