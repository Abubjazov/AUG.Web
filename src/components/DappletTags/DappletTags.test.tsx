import { fireEvent, render, screen } from '@testing-library/react'
import * as asyncActions from 'store/asyncThunks/userData'
import AddUserTagModalContent from 'components/AddUserTagModalContent'
import { mockCommunityTags, mockUserTags } from 'mockData/mockData'
import { mockedReduxProvider as Provider } from 'mockData/mockedReduxProvider'
import * as actions from 'store/slices/layoutSlice'
import { EDappletOperation } from 'store/slices/userDataSlice'

import DappletTags from './DappletTags'

describe('DappletTags', () => {
  test('should render DappletTags default & dappletState is false', () => {
    const { asFragment } = render(
      <Provider>
        <DappletTags
          dappletId={'ECNk2nNngwGXouvMpjWt'}
          dappletState={false}
          dappletUserTags={mockUserTags.slice(0, 3)}
          dappletCommunityTags={mockCommunityTags.slice(0, 2)}
        />
      </Provider>,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test('should render DappletTags default & dappletState is true', () => {
    const { asFragment } = render(
      <Provider>
        <DappletTags
          dappletId={'ECNk2nNngwGXouvMpjWt'}
          dappletState={true}
          dappletUserTags={mockUserTags.slice(0, 3)}
          dappletCommunityTags={mockCommunityTags.slice(0, 2)}
        />
      </Provider>,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test('should render MyLists when the "dappletOperationGoing"', () => {
    const { asFragment } = render(
      <Provider>
        <DappletTags
          dappletId={'ECNk2nNngwGXouvMpjWt'}
          dappletState={true}
          dappletUserTags={mockUserTags.slice(0, 3)}
          dappletCommunityTags={mockCommunityTags.slice(0, 2)}
          dappletOperationGoing={[
            {
              dappletId: 'ECNk2nNngwGXouvMpjWt',
              userTagId: 'JW3UFtZ5HgATcwldsJ1T',
              operation: EDappletOperation.REMOVE_USER_TAG,
            },
          ]}
        />
      </Provider>,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test('should call installDapplet function on delete tag button click', () => {
    const mockedRemoveUserTagFromDapplet = vi
      .spyOn(asyncActions, 'removeUserTagFromDapplet')
      .mockImplementation(() => vi.fn())

    render(
      <Provider>
        <DappletTags
          dappletId={'ECNk2nNngwGXouvMpjWt'}
          dappletState={true}
          dappletUserTags={mockUserTags.slice(0, 3)}
          dappletCommunityTags={mockCommunityTags.slice(0, 2)}
        />
      </Provider>,
    )

    fireEvent.click(
      screen.getByTestId('smart-tag-cross-button-JW3UFtZ5HgATcwldsJ1T'),
    )

    expect(mockedRemoveUserTagFromDapplet).toHaveBeenCalledTimes(1)
    expect(mockedRemoveUserTagFromDapplet).toHaveBeenCalledWith({
      dappletId: 'ECNk2nNngwGXouvMpjWt',
      userTagId: 'JW3UFtZ5HgATcwldsJ1T',
    })
  })

  test('should render DappletTags then windowInnerWidth <= 880 & dappletState is false', () => {
    window.innerWidth = 880

    const { asFragment } = render(
      <Provider>
        <DappletTags
          dappletId={'ECNk2nNngwGXouvMpjWt'}
          dappletState={false}
          dappletUserTags={mockUserTags.slice(0, 2)}
          dappletCommunityTags={mockCommunityTags.slice(0, 1)}
        />
      </Provider>,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test('should render DappletTags then windowInnerWidth <= 880 & dappletState is true', () => {
    window.innerWidth = 880

    const { asFragment } = render(
      <Provider>
        <DappletTags
          dappletId={'ECNk2nNngwGXouvMpjWt'}
          dappletState={true}
          dappletUserTags={mockUserTags.slice(0, 2)}
          dappletCommunityTags={mockCommunityTags.slice(0, 1)}
        />
      </Provider>,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test('should call setModalInner & setModalState functions on tag click then windowInnerWidth <= 880', () => {
    window.innerWidth = 880

    const mockedSetModalInner = vi.spyOn(actions, 'setModalInner')
    const mockedSetModalState = vi.spyOn(actions, 'setModalState')

    render(
      <Provider>
        <DappletTags
          dappletId={'ECNk2nNngwGXouvMpjWt'}
          dappletState={true}
          dappletUserTags={mockUserTags.slice(0, 2)}
          dappletCommunityTags={mockCommunityTags.slice(0, 1)}
        />
      </Provider>,
    )

    fireEvent.click(screen.getByTestId('add-tag-button'))

    expect(mockedSetModalInner).toHaveBeenCalledTimes(1)
    expect(mockedSetModalInner).toHaveBeenCalledWith(
      <AddUserTagModalContent dappletId={'ECNk2nNngwGXouvMpjWt'} />,
    )

    expect(mockedSetModalState).toHaveBeenCalledTimes(1)
    expect(mockedSetModalState).toHaveBeenCalledWith(true)
  })
})
