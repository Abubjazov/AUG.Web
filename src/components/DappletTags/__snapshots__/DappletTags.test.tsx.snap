// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DappletTags > should render DappletTags default 1`] = `
<DocumentFragment>
  <div
    class="_root_dea2c3 "
  />
</DocumentFragment>
`;

exports[`DappletTags > should render DappletTags then windowInnerWidth <= 880 & dappletState is false 1`] = `
<DocumentFragment>
  <div
    class="_root_dea2c3 "
  />
</DocumentFragment>
`;

exports[`DappletTags > should render DappletTags then windowInnerWidth <= 880 & dappletState is true 1`] = `
<DocumentFragment>
  <div
    class="_root_dea2c3 "
  >
    <button
      class="_button_dea2c3"
      data-testid="add-tag-button"
      type="button"
    >
      +
    </button>
  </div>
</DocumentFragment>
`;
